src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -370,68 +390,85 @@ public void Reset()
+                        {
+                            if (readResult.Buffer.Length != 0)
+                            {
+                                ThrowUnexpectedState();
+                            }
 
-                        // Headers have already been written and there is no other content to write
-                        flushResult = await _frameWriter.FlushAsync(outputAborter: null, cancellationToken: default);
-                    }
-                    else
-                    {
-                        var endStream = readResult.IsCompleted;
-                        if (endStream)
+                            // Headers have already been written and there is no other content to write
+                            flushResult = await _frameWriter.FlushAsync(outputAborter: null, cancellationToken: default);
+                        }
+                        else
                         {
-                            _stream.DecrementActiveClientStreamCount();
+                            var endStream = readResult.IsCompleted;
+                            if (endStream)
+                            {
+                                _stream.DecrementActiveClientStreamCount();
+                            }
+                            flushResult = await _frameWriter.WriteDataAsync(StreamId, _flowControl, readResult.Buffer, endStream, forceFlush: true);
                         }
-                        flushResult = await _frameWriter.WriteDataAsync(StreamId, _flowControl, readResult.Buffer, endStream, forceFlush: true);
-                    }
 
-                    _pipeReader.AdvanceTo(readResult.Buffer.End);
-                } while (!readResult.IsCompleted);
-            }
-            catch (Exception ex)
-            {
-                _log.LogCritical(ex, nameof(Http2OutputProducer) + "." + nameof(ProcessDataWrites) + " observed an unexpected exception.");
-            }
+                        _pipeReader.AdvanceTo(readResult.Buffer.End);
+                    } while (!readResult.IsCompleted);
+                }
+                catch (Exception ex)
+                {
+                    _log.LogCritical(ex, nameof(Http2OutputProducer) + "." + nameof(ProcessDataWrites) + " observed an unexpected exception.");
+                }
 
-            _pipeReader.Complete();
+                _pipeReader.Complete();
 
-            return flushResult;
+                // Signal via WriteStreamSuffixAsync to the stream that output has finished.
+                // Stream state will move to RequestProcessingStatus.ResponseCompleted
+                _responseCompleteTaskSource.SetResult(flushResult);
+
+                if (readResult.IsCompleted)
+                {
+                    // Successfully read all data. Wait here for the stream to be reset.
+                    await new ValueTask(_resetAwaitable, _resetAwaitable.Version);
+                    _resetAwaitable.Reset();
```

Comment:
If Dispose is called, you would be resetting a field in an already disposed object.

Reply:
This object isn't disposed, it'll just reset the awaitable which is harmless.